bentoml/BentoML
# feat: test bento bundle

## What does this PR address?

This PR adds the `bentoml test` command which enables testing a Bento for making sure it is correctly packaged.
In order to use `bentoml test` the Bento builder must supply input/output example(s), this will be done in the `bentofile.yaml` in a dedicated section called `endpoints_tests`. 
Once specified, the Bento builder can call `bentoml test <bento-tag>` to test the specified bento, or alternatively call `bentoml build --test` which will build and test in one call (still not implemented).

The testing itself will invoke the `bentoml containerize` command to create an image, will launch a docker container from that image and then will interact with the docker through HTTP request to test the endpoints. This will cover all the critical functionality of Bentos; creating an environment from the Bento specification and the service's endpoints.


## Example

Imagine I have the below, dummy, scikit-learn scaler model serving:
```python
# init
model_name = 'test-scaler'

# load model's metadata
scaler_meta = bentoml.sklearn.get(f'{model_name}:latest')

# create service
scaler_runner = scaler_meta.to_runner()
scaler_service = bentoml.Service(model_name, runners=[scaler_runner])

# create endpoint for the service
@scaler_service.api(input=PandasDataFrame(), output=NumpyNdarray())
def my_predict_endpoint(input_df: pd.DataFrame) -> pd.DataFrame:
    res = scaler_runner.run(input_df)
    return res
```

The altered `Bentofile.yaml` will look as follows (note the new `endpoints_tests` section):
```yaml
service: "serve_model:scaler_service"  # Same as the argument passed to `bentoml serve`
labels:
  owner: bentoml-team
  stage: dev
include:
  - "*.py"  # A pattern for matching which files to include in the bento
python:
  packages: # Additional pip packages required by the service
    - scikit-learn
    - pandas

## TESTING
endpoints_tests:
  my_predict_endpoint:
    - input: [[1, 2, 3, 4, 5]]
      output: [[-1.4638501094227998,-1.4638501094227998,-0.8783100656536799,-0.29277002188455997,-0.8783100656536799]]
    - input: [[1, 2, 3, 4, 5]]
```

As you can see, I made 2 tests for `my_predict_endpoint` endpoint:
1. one that takes both input and output.
2. one that takes only an input.

Now, when building a Bento, using `bentoml build`, we will get a Bento tag, somewhat similar to `test-scaler:e45pn7bm7cr4rdsd`.


Now we can test our newly created Bento, using the command `bentoml test test-scaler:e45pn7bm7cr4rdsd`, which will do the following:
1. Create a docker image from the given Bento (using the `containerize` command under the hood).
2. Run a docker container from the newly created docker image.
3. Run the 2 tests, specified in the `Bentofile.yaml`, against the `my_predict_endpoint` endpoint:
i. The 1st test, which takes the input, send it to the endpoint, and check if the received result equals to the expected output specified in the test (`[[-1.4638501094227998, ...]]`).
ii. The 2nd test, which takes the input, send it to the endpoint, and just check that the service didn't throw and error (unlike the 1st test, it doesnt do any checking on the responded output). 



## Issue
https://github.com/bentoml/BentoML/issues/2967

## Before submitting:



- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
  those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

## Who can help review?
@aarnphm
@parano

Feel free to tag members/contributors who can help review your PR.
